src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -844,6 +965,12 @@ SyntaxAnalyzerResultPtr SyntaxAnalyzer::analyzeSelect(
         /// Remove duplicate items from ORDER BY.
         optimizeOrderBy(select_query);
 
+        /// Remove duplicate ORDER BY from subqueries.
+        optimizeDuplicateOrderBy(query, context);
+
+        /// Remove duplicate DISTINCT from subqueries.
+        optimizeDuplicateDistinct(query);
```

Comment:
This does not work for the following case:

```sql
SELECT DISTINCT number
FROM 
(
    SELECT DISTINCT number
    FROM remote('127.{2,3}', system.numbers)
    LIMIT 1
    SETTINGS distributed_group_by_no_merge = 1
)

┌─number─┐
│      0 │
└────────┘
┌─number─┐
│      0 │
└────────┘
```

```sql
SELECT DISTINCT number
FROM 
(
    SELECT number
    FROM remote('127.{2,3}', system.numbers)
    LIMIT 1
    SETTINGS distributed_group_by_no_merge = 1
)

┌─number─┐
│      0 │
└────────┘
```

Both should return only one row